packages/controllers/src/permissions/rpc-methods/requestPermissions.ts

```diff
@@ -0,0 +1,64 @@
+import type { PermittedHandlerExport } from '@metamask/rpc-methods/types';
+import type {
+  JsonRpcEngineEndCallback,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from 'json-rpc-engine';
+import { nanoid } from 'nanoid';
+import { MethodNames } from '../utils';
+
+import { invalidParams } from '../errors';
+import type { PermissionConstraint, RequestedPermissions } from '../Permission';
+
+export const requestPermissionsHandler: PermittedHandlerExport<
+  RequestPermissionsHooks,
+  [RequestedPermissions],
+  PermissionConstraint[]
+> = {
+  methodNames: [MethodNames.requestPermissions],
+  implementation: requestPermissionsImplementation,
+  methodDescription: 'Request permissions.',
+  hookNames: {
+    requestPermissions: true,
+  },
+};
+
+type RequestPermissions = (
+  requestedPermissions: RequestedPermissions,
+  id: string,
+) => Promise<
+  [Record<string, PermissionConstraint>, { id: string; origin: string }]
+>;
+
+export type RequestPermissionsHooks = {
+  requestPermissions: RequestPermissions;
+};
+
+async function requestPermissionsImplementation(
+  req: JsonRpcRequest<[RequestedPermissions]>,
+  res: PendingJsonRpcResponse<PermissionConstraint[]>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  { requestPermissions }: RequestPermissionsHooks,
+): Promise<void> {
+  if (
+    !Array.isArray(req.params) ||
+    !req.params[0] ||
+    typeof req.params[0] !== 'object' ||
+    Array.isArray(req.params[0])
+  ) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  const id =
+    typeof req.id === 'number' || req.id ? req.id.toString() : nanoid();
```

Comment:
This is the JSON-RPC request id, so that should (and in our case, is) set by the JSON-RPC processing pipeline. Our IDs are guaranteed to be unique within our application as well.

Reply:
Do we need to set our own ID, in that case? We could instead throw if it's missing.

I don't love the idea of having fallback branches that don't have any real purpose. It's more work to test them, and they might introduce complicated bugs. e.g. it might hide that the `id` is missing when it shouldn't have been, or it might confuse matters by introducing a new ID format (if `nanoid` isn't used elsewhere).

Reply:
Fair enough: 118bbc9